File: valid-parentheses/valid-parentheses.js
let isValid = function(s) {
    const endToStart = {')': '(', ']': '[', '}': '{'}
    let stack = []

    const crumbs = s.split('')
    for (let i = 0; i < crumbs.length; i++) {
        let current = crumbs[i]
        if (stack.length && current in endToStart) {
            if (stack[stack.length - 1] === endToStart[current]) {
                stack.pop()

                continue
            }

            return false
        }

        stack.push(current)
    }

    return !stack.length
};

File: valid-parentheses/valid-parentheses.py
def isValid(s):
    endToStart = {')': '(', ']': '[', '}': '{'}
    stack = []

    for current in s:
        if current in endToStart:
            if stack and stack[-1] == endToStart[current]:
                stack.pop()

                continue

            return False

        stack.append(current)

    return len(stack) < 1